daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala

```diff
@@ -30,13 +30,99 @@ sealed trait SValue {
   /** Convert a speedy-value to a value which may not be correctly normalized.
     * And so the resulting value should not be serialized.
     */
-  def toUnnormalizedValue: V[V.ContractId] = SValue.toValue(this)
+  def toUnnormalizedValue: V[V.ContractId] = {
+    toValue(
+      disallowGenMapAtVersion = None,
+      eraseType = false,
+    )
+  }
 
   /** Convert a speedy-value to a value normalized according to the LF version.
     */
   def toNormalizedValue(version: TransactionVersion): V[V.ContractId] = {
-    val v = SValue.toValue(this)
-    Util.assertNormalizeValue(v, version) //TODO: avoid separate pass; inline norm into toValue
+    import Ordering.Implicits.infixOrderingOps
+    toValue(
+      disallowGenMapAtVersion =
+        if (version >= TransactionVersion.minGenMap) None else Some(version),
+      eraseType = version >= TransactionVersion.minTypeErasure,
+    )
+  }
+
+  private def toValue(
+      disallowGenMapAtVersion: Option[TransactionVersion],
+      eraseType: Boolean,
```

Comment:
```suggestion
      normalize: Boolean,
```
maybe a bit clearer if we add more normalization steps in the future?